### PR TITLE
Make cycle reported by toposort shorter.

### PR DESCRIPTION
**Pull request by eriknw**
This is an alternative to #9057.  Previously, the cycle reported by toposort could contain extra dependencies of nodes that are part of the cycle.  Now we use a simple heuristic to greedily construct a short cycle that only contains nodes in the cycle, but there is no guarantee that this is the shortest cycle.

- [X] Tests added / passed
- [X] Passes `pre-commit run --all-files`

I'm a little sick and not 100% myself, so I'm not as confident that there isn't a better/cleaner way to do this as I would normally be.  Hopefully the tests are reliable.  These tests are how I tested this locally, and before this PR, the cycles were of length 9.

### CODE REVIEW

File: dask/core.py [hunk 1]
```diff
@@ -404,11 +404,33 @@ def _toposort(dsk, keys=None, returncycle=False, dependencies=None):
...
+                        # Let's report only the nodes that directly participate in the cycle.
+                        # We use `priorities` below to greedily construct a short cycle.
+                        # Shorter cycles may exist.
+                        priorities = {}
+                        prev = nodes[-1]
+                        # Give priority to nodes that were seen earlier.
                         while nodes[-1] != nxt:
-                            cycle.append(nodes.pop())
-                        cycle.append(nodes.pop())
+                            priorities[nodes.pop()] = -len(priorities)
```
Comment (fengsp) [inline dask/core.py:414]: I do not understand `priorities` here. Hash in Python 3 random -> dependency_set random order -> stack random order -> priorities random, so we are just picking a random `prev dep`?

**Comment by jsignell** [comment]
@eriknw sorry this one got lost in the shuffle. If you are happy with this change I can go ahead and merge.

File: dask/core.py (hunk 1 above)
Comment (eriknw) [reply dask/core.py:414]: `priorities` is used as the key in the `min` call below to construct shorter chains.  This let's us take bigger steps when there are multiple options when "walking backwards", and bigger steps results in fewer steps (hopefully).

To clarify, we use `priorities.__getitem__` as the key function for `min`.  For example, `key={1: 10, 2:20}.__getitem__` is a function that maps `1` to `10` and `2` to `20`.

**Comment by eriknw** [comment]
No worries @jsignell, and thanks for circling back around and checking with me.

Yeah, I think this is good to go in.  It's kinda hard (i.e., annoying) to make a good test for this, but I think the current test I added is probably good enough.

**Comment by jsignell** [comment]
Thanks @eriknw for this work!

**Pull request merged**